2021-12-27-getting-started-with-purenix.md

```diff
@@ -0,0 +1,217 @@
+------------------------------------------------------
+title: Getting Started with PureNix
+summary: How to setup an environment for playing around with PureNix
+tags: haskell, nixos, purescript
+draft: false
+------------------------------------------------------
+
+*This post is the first post in a
+[series about PureNix](./2021-12-26-purenix).*
+
+[PureNix](https://github.com/purenix-org/purenix) is a Nix backend for
+[PureScript](https://www.purescript.org/).  PureNix allows you to write
+PureScript code and transpile it to Nix code.
+
+This post gives an overview of PureScript and PureNix. It then explains how to
+get started using PureNix.
+
+## PureScript
+
+[PureScript](https://en.wikipedia.org/wiki/PureScript) is a strongly-typed,
+pure functional programming language. Here's a small example of a function
+written in PureScript.  If you know Haskell, this should look very familiar:
+
+```purescript
+module Main where
+
+import Prelude
+
+type FilePath = String
+
+subdirectory :: FilePath -> FilePath -> FilePath
+subdirectory p1 p2 = p1 <> "/" <> p2
+```
+
+The PureScript compiler compiles code like this to JavaScript.  PureScript ends
+up being a Haskell-like language that you can use to write frontend code.
+
+One of the main build tools for PureScript is
+[Spago](https://github.com/purescript/spago).  Spago is somewhat similar to
+Haskell's `stack` or `cabal`, Rust's `cargo`, etc.
+
+One interesting feature of the PureScript compiler and Spago is that there is
+good support for switching out the backend code generator.  The code generator
+built into the main PureScript compiler is for JavaScript, but there are
+alternative code generators for other languages.  PureNix is an alternative
+backend code generator for Nix.
+
+## PureNix
+
+PureNix works together with PureScript and Spago to compile PureScript code
+into Nix code.  Here's a small example of PureScript code, and the Nix code it
+gets compiled to:
+
+```purescript
+module Main where
+
+greeting :: String
+greeting = "Hello, world!"
+
+identity :: forall a. a -> a
+identity a = a
+
+data Maybe a = Nothing | Just a
+
+fromMaybe :: forall a. a -> Maybe a -> a
+fromMaybe a Nothing = a
+fromMaybe _ (Just a) = a
+```
+
+PureNix compiles this PureScript code to the following Nix code[^1]:
```

Comment:
FWIW, footnotes can be named, and the body doesn't have to be at the bottom: https://github.com/jonascarpay/jonascarpay.github.io/blob/master/posts/2021-05-16-parsing-programming-language.md?plain=1#L147

Reply:
Oh thanks!  That's much nicer!  I fixed this in 78c9dd8.